man/man7/zmq.7
.TH zmq 7 "" "(c)2007-2009 FastMQ Inc." "0MQ User Manuals"
.SH NAME
0MQ \- lightweight messaging kernel
.SH DESCRIPTION
0MQ is an extensible messaging framework. It allows you to send messages though
the network at hight data rates and low latencies. 
.SH MULTI-PROTOCOL SUPPORT
0MQ allows for different transport-level and MOM-level protocols.
Following section describes how to specify where the entities are
located on the network (see
.IR api_thread_t::create_exchange
and
.IR api_thread_t::create_queue
methods).
.IP "\fB0MQ backend protocol over TCP/IP\fP"
.RS
.TP 10
.I Format:
zmq.tcp://network-interface:port
.TP 
Examples: zmq.tcp://eth0:5555 or zmq.tcp://192.168.1.1:5555
.RE
.IP "\fBSCTP protocol\fP"
.RS
.TP 10
.I Format:
sctp://network-interface:port
.TP 
Examples: sctp://eth0:5555 or sctp://192.168.1.2:5555
.RE
.IP "\fBAMQP protocol\fP"
.RS
0MQ implements only client side of AMQP at the moment. Thus it allows you to
connect to an existing AMQP broker, however, it doesn't allow your
0MQ application to play a role of AMQP broker. To define location of AMQP broker
to connect to specify its location in config file and pass it to
.IR zmq_server
on the start-up.
.TP 10
.I Format:
amqp://broker-host:port
.TP 
Example:  amqp://192.168.1.3:5672 or amqp://server001:5672
.RE
.IP "\fB0MQ backend protocol over PGM\fP"
.RS
.TP 10
.I Format:
zmq.pgm://network-interface;multicast-group:port
.BR
zmq.pgm://udp:network-interface;multicast-group:port
.TP
The latter version causes PGM packets to be encapsulated in UDP packets.
.TP
Example:  zmq.pgm://eth0;226.0.0.1:7501
.RE
.SH WEB PAGE
Documentation related to 0MQ project can be found on \fBwww.zeromq.org\fP.
.SH MAILING LIST
Mailing list for 0MQ project is \fB<zeromq-dev@lists.zeromq.org>\fP
.SH AUTHOR
Pavol Malosek <malosek at fastmq dot com>
.SH "SEE ALSO"
.BR zmq_server (1),
.BR zmq::locator_t (3),
.BR zmq::api_thread_t (3),
.BR zmq::io_thread_t (3),
.BR zmq::dispatcher_t (3),
.BR zmq::message_t (3),
.BR zmq::select_thread_t (3),
.BR zmq::poll_thread_t (3),
.BR zmq::epoll_thread_t (3),
.BR zmq::devpoll_thread_t (3),
.BR zmq::kqueue_thread_t (3),
.BR zmq::set_error_handler (3)



ChangeLog
Version 0.6

ZMQ-73: Redesign MT throughput test
ZMQ-102: Move load-balancing functionality from 0.3.3 branch to the trunk
ZMQ-136: Check I/O threads for rm_fd(x) followed by add_fd(x) behavior
ZMQ-138: Avoid exporting 0MQ symbols from libs using 0MQ on Windows
ZMQ-147: Python app hang-up, unresponsive to Ctrl+C
ZMQ-164: Add reconnection functionality to PGM
ZMQ-166: Slow down the reconnection process
ZMQ-176: Update examples to reflect the auto-reconnect feature
ZMQ-179: Implement OS-agnostic time measurement
ZMQ-180: What happens if queue is full & disconnection occurs?
ZMQ-183: Change OpenPGM to PGM in build options
ZMQ-185: Move msg_store to the trunk
ZMQ-186: Write an example to demonstrate load-balancing
ZMQ-187: Declaring same object (exchange, queue) twice doesn't fail
ZMQ-188: Schemas used by global object URIs don't conform to RFC3986
ZMQ-190: Compile libdnzmq with /clr support (no old syntax)
ZMQ-193: Reconnect on Solaris platform not working
ZMQ-194: Let sender use RDTSC mechanism for command throttling in MSVC build
ZMQ-195: Failure during fg
ZMQ-196: System out of sockets after number of reconnections
ZMQ-197: Offload signal handling from I/O threads
ZMQ-198: Timers can never expire
ZMQ-199: Make receive in libczmq zero-copy
ZMQ-200: Virtualise global locator
ZMQ-202: Check out SCTP support on QNX
ZMQ-203: Add auto-reconnect to AMQP engine
ZMQ-205: Check handling of async errors on a socket
ZMQ-206: Make zmq_server not to drop the endpoint info on service disconnect
ZMQ-207: zmq_server should be able to read endpoint info from XML file
ZMQ-210: Switch perf tests to load-balancing mode
ZMQ-211: Port Data Dam functionality to Win32
ZMQ-212: Integrate add and inc in atomic_counter into a single function
ZMQ-213: When 0MQ infrastructure is terminated, sockets are not closed correctly
ZMQ-215: Java on Win32 platform not stopping in case of assertion
ZMQ-217: zmq_server can be crashed by incorect input data
ZMQ-218: C# integration not working anymore
ZMQ-219: Change .NET extesnsion to use CamelCase
ZMQ-220: Check whether individual APIs conforms to API guidelines document
ZMQ-221: Allow to create global object w/o specifying the location
ZMQ-222: SPARC code for atomic_counter broken
ZMQ-223: Rename czmq/jzmq/dnzmq etc. to simple "zmq"
ZMQ-225: Application fails when it's connected from telnet

Version 0.5

ZMQ-17: look at pgm_transport_set_recv_only and ...send_only
ZMQ-26: Allow for "any" network interface
ZMQ-31: Implement SCTP support
ZMQ-45: Normalise API vs. I/O thread
ZMQ-78: Make connection establishment asynchronous
ZMQ-79: Refactor handling of broken connections
ZMQ-100: Move flow control and queue limits from 0.3.3 branch to the trunk
ZMQ-101: Move reconnection functionality from 0.3.3 branch to the trunk
ZMQ-112: build via cmake (makefile generator)
ZMQ-115: Identify SPARC V9 microarc and/or availability of CAS instruction
ZMQ-116: Implement .NET extension
ZMQ-129: Make engine_source part of the polling object
ZMQ-139: /dev/poll introduces latency on Solaris
ZMQ-144: Investigate latency impact of using eventfd instead of socketpair on Linux
ZMQ-146: Check whether i386 and x86_64 in atomic_ptr can be unified
ZMQ-148: Change watermarks from int to uint64_t
ZMQ-149: Virtualise creation of 0MQ connections
ZMQ-150: Negative numbers in unsigned type in atomic_counter
ZMQ-151: Modify build system to build SCTP extension
ZMQ-152: Merge BP/PGM functionality into the trunk
ZMQ-153: compiler warnings on solaris
ZMQ-155: Build libzmq with sctp support witch cmake
ZMQ-160: Get rid of create functions in engines
ZMQ-162: CMakeLists.txt files not updated with latest changes
ZMQ-165: Implement reconnection notifications
ZMQ-169: Return queue id from receive function in libpyzmq
ZMQ-171: Add AMQP support to 0MQ
ZMQ-173: Throughput test should accept 2 interface specifications
ZMQ-174: When application is killed on Win32, zmq_server fails
ZMQ-175: Chatroom example doesn't pass messages after reconnection
ZMQ-177: Update language extensions to take advantage of auto-reconnect
ZMQ-181: Rafactor the code to support multiple notification types

Version 0.4

ZMQ-21: Clean up the command passing code
ZMQ-36: autogen/configure fails on Solaris
ZMQ-39: Check whether using CAS for btsr and izte on x86 platform improves latency
ZMQ-40: Check autogen on FreeBSD
ZMQ-41: Check autogen on OS X
ZMQ-43: Use +m instead of memory in atomic operations
ZMQ-46: Get rid of administrative pseudo-thread
ZMQ-63: -mcpu=v9 CPPFLAG should be used only at sparc achitecture
ZMQ-64: gcc 3.4.3 on opensolaris complains about -mcpu flag
ZMQ-65: No error report when host name resolving fails
ZMQ-66: Several files use atoi without including stdlib.h
ZMQ-69: Make stdint.hpp hand-written rather than generated
ZMQ-71: Virtualise thread management
ZMQ-72: Remove platform.hpp
ZMQ-74: Implement win32 version of tcp_socket and tcp_listener
ZMQ-75: Virtualise mutexes
ZMQ-76: Implement win32 version of ysemaphore_t
ZMQ-80: Virtualise polling mechanisms
ZMQ-81: Implement 'select' polling mechanism
ZMQ-82: Rename i_context to i_thread
ZMQ-83: "Resource temporarily unavailable" when running on 266MHz box
ZMQ-84: zmq_server stops to respond
ZMQ-89: Make i_pollable derived from i_engine
ZMQ-94: Report error when malloc fails
ZMQ-95: Add /dev/poll method of polling
ZMQ-96: Add support for kqueue method of polling
ZMQ-97: OpenBSD port
ZMQ-98: QNX port
ZMQ-99: use pyexec_LTLIBRARIES primary for libpyzmq
ZMQ-103: Use non blocking sockets
ZMQ-106: memory leak when sending message from Java
ZMQ-111: Get rid of speculative operations
ZMQ-113: Windows compiler (_MSC_VER 1310) doesn't support strncpy_s function
ZMQ-117: select_thread doesn't implement rm_fd function
ZMQ-119: select thread fails on shudown
ZMQ-121: Fix all I/O threads to handle rm_fd correctly
ZMQ-124: include gaurd is wrong
ZMQ-126: Refactor I/O threads implementation to avoid duplication of common functionality.
ZMQ-131: select_thread fails on Windows

libjzmq/Zmq.cpp
/*
    Copyright (c) 2007-2009 FastMQ Inc.

    This file is part of 0MQ.

    0MQ is free software; you can redistribute it and/or modify it under
    the terms of the Lesser GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    0MQ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    Lesser GNU General Public License for more details.

    You should have received a copy of the Lesser GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

//  The header file generated by JNI from Zmq.java source file.
#include "Zmq.h"

#include <stdio.h>
#include <assert.h>

#include <zmq.hpp>

static jfieldID context_fid = NULL;

struct context_t
{
    //  Cache for JNI static objects.
    jclass inbound_data_class;
    jfieldID queue_fid;
    jfieldID type_fid;
    jfieldID message_fid;

    //  0MQ infrastructure.
    zmq::locator_t *locator;
    zmq::dispatcher_t *dispatcher;
    zmq::i_thread *io_thread;
    zmq::api_thread_t *api_thread;
};

JNIEXPORT void JNICALL Java_Zmq_construct (JNIEnv *env, jobject obj,
    jstring host)
{
    //  Cache context field ID.
    jclass cls = env->GetObjectClass (obj);
    context_fid = env->GetFieldID (cls, "context", "J");
    assert (context_fid);

    //  Get the hostname.
    assert (host);
    char *c_host = (char*) env->GetStringUTFChars (host, 0);
    assert (c_host);

    //  Create the context.
    context_t *context = new context_t;
    assert (context);
    context->dispatcher = new zmq::dispatcher_t (2);
    assert (context->dispatcher);
    context->locator = new zmq::locator_t (c_host);
    assert (context->locator);
    context->io_thread = zmq::select_thread_t::create (context->dispatcher);
    assert (context->io_thread);
    context->api_thread = zmq::api_thread_t::create (context->dispatcher,
        context->locator);
    assert (context->api_thread);

    //  Cache Zmq.InboundData class.
    jclass idc = env->FindClass ("Zmq$InboundData");
    assert (idc);
    context->inbound_data_class = (jclass) env->NewGlobalRef (idc);
    assert (context->inbound_data_class);
    env->DeleteLocalRef (idc);

    //  Cache fieldIDs for InboundData member variables.
    context->queue_fid = env->GetFieldID (context->inbound_data_class,
        "queue", "I");
    assert (context->queue_fid);
    context->type_fid = env->GetFieldID (context->inbound_data_class,
        "type", "I");
    assert (context->type_fid);
    context->message_fid = env->GetFieldID (context->inbound_data_class,
        "message", "[B");
    assert (context->message_fid);

    //  Clean-up.
    env->ReleaseStringUTFChars (host, c_host);

    //  Set the context.
    env->SetLongField (obj, context_fid, (jlong) context);
}

JNIEXPORT void JNICALL Java_Zmq_finalize (JNIEnv *env, jobject obj)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Free JNI cached JNI objects.
    env->DeleteGlobalRef ((jobject) context->inbound_data_class);

    //  Deallocate the 0MQ infrastructure.
    delete context->locator;
    delete context->dispatcher;

    delete context;
}

JNIEXPORT void JNICALL Java_Zmq_mask (JNIEnv *env, jobject obj,
    int notifications)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);
    
    //  Forward the call.
    context->api_thread->mask (notifications);
}

JNIEXPORT jint JNICALL Java_Zmq_createExchange (JNIEnv *env, jobject obj,
    jstring name, jint scope, jstring location, jint style)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Get the exchange name.
    assert (name);
    char *c_name = (char*) env->GetStringUTFChars (name, 0);
    assert (c_name);

    //  Get the location.
    char *c_location = NULL;
    if (location)
        c_location = (char*) env->GetStringUTFChars (location, 0);

    //  Forward the call.
    jint eid = context->api_thread->create_exchange (c_name,
        (zmq::scope_t) scope, c_location, context->io_thread, 1,
        &context->io_thread, (zmq::style_t) style);

    //  Clean-up.
    env->ReleaseStringUTFChars (name, c_name);
    if (c_location)
        env->ReleaseStringUTFChars (location, c_location);

    return eid;
}

JNIEXPORT jint JNICALL Java_Zmq_createQueue (JNIEnv *env, jobject obj,
    jstring name, jint scope, jstring location, jlong hwm, jlong lwm, 
    jlong swap)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Get the queue name.
    assert (name);
    char *c_name = (char*) env->GetStringUTFChars (name, 0);
    assert (c_name);

    //  Get NIC name.
    char *c_location = NULL;
    if (location)
        c_location = (char*) env->GetStringUTFChars (location, 0);

    //  Forward the call.
    jint qid = context->api_thread->create_queue (c_name, (zmq::scope_t) scope,
        c_location, context->io_thread, 1, &context->io_thread,
        (int64_t) hwm, (int64_t) lwm, (int64_t) swap);

    //  Clean-up.
    env->ReleaseStringUTFChars (name, c_name);
    if (c_location)
        env->ReleaseStringUTFChars (location, c_location);

    return qid;
}

JNIEXPORT void JNICALL Java_Zmq_bind (JNIEnv *env, jobject obj,
    jstring exchangeName, jstring queueName, jstring exchangeOptions,
    jstring queueOptions)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Get the exchange name.
    assert (exchangeName);
    char *c_exchange_name = (char*) env->GetStringUTFChars (exchangeName, 0);
    assert (c_exchange_name);

    //  Get the queue name.
    assert (queueName);
    char *c_queue_name = (char*) env->GetStringUTFChars (queueName, 0);
    assert (c_queue_name);

    //  Get exchange options.
    char *c_exchange_options = NULL;
    if (exchangeOptions)
        c_exchange_options = (char*) env->GetStringUTFChars (
            exchangeOptions, 0);

    //  Get queue options.
    
    char *c_queue_options = NULL;
    if (queueOptions)
        c_queue_options = (char*) env->GetStringUTFChars (queueOptions, 0);

    //  Forward the call.
    context->api_thread->bind (c_exchange_name, c_queue_name,
        context->io_thread, context->io_thread, c_exchange_options,
        c_queue_options);

    //  Clean-up.
    env->ReleaseStringUTFChars (exchangeName, c_exchange_name);
    env->ReleaseStringUTFChars (queueName, c_queue_name);
    if (c_exchange_options)
        env->ReleaseStringUTFChars (exchangeOptions, c_exchange_options);
    if (c_queue_options)
        env->ReleaseStringUTFChars (queueOptions, c_queue_options);
}

JNIEXPORT jboolean JNICALL Java_Zmq_send (JNIEnv *env, jobject obj,
    jint exchange, jbyteArray message, jboolean block)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Get the data from the bytearray.
    jsize size = env->GetArrayLength (message); 
    jbyte *data = env->GetByteArrayElements (message, 0);

    //  Create the message.
    zmq::message_t msg (size);
    memcpy (msg.data (), data, size);

    //  Release the bytearray.
    env->ReleaseByteArrayElements (message, data, 0);

    //  Send the message.
    return (jboolean) context->api_thread->send (exchange, msg, block);
}

JNIEXPORT jobject JNICALL Java_Zmq_receive (JNIEnv *env, jobject obj,
    jboolean block)
{
    //  Get the context.
    context_t *context = (context_t*) env->GetLongField (obj, context_fid);
    assert (context);

    //  Allocate new InboundData object; no constructor is called!
    //  We're doing the allocation before the receive so that delivery of
    //  the message is not slowed down by the allocation.
    jobject result = env->AllocObject (context->inbound_data_class);
    assert (result);

    //  Get new message.
    zmq::message_t msg;
    int qid = context->api_thread->receive (&msg, block);

    //  Fill in the InboundData object.
    env->SetIntField (result, context->queue_fid, (jint) qid);
    env->SetIntField (result, context->type_fid, (jint) msg.type ());
    jbyteArray data = env->NewByteArray (msg.size ());
    assert (data);
    env->SetByteArrayRegion (data, 0, msg.size (), (jbyte*) msg.data ());
    env->SetObjectField (result, context->message_fid, data);
    env->DeleteLocalRef (data);

    return result;
}
